src/client.rs

```diff
@@ -0,0 +1,308 @@
+            .send()
+            .await
+            .map_err(|e| map_reqwest_error(uri, "POST", e))
+    }
+
+    pub async fn get_actor(&self, uri: &str) -> Result<Actor> {
+        match self.json_get(uri).await {
+            Ok(actor) => Ok(actor),
+
+            Err(Error::FailedRequest { status, .. }) if status == StatusCode::NOT_FOUND => {
+                info!(%uri, "failed to fetch actor");
+                Err(Error::StatusAndMessage {
+                    status: StatusCode::NOT_FOUND,
+                    message: "failed to fetch actor",
+                })
+            }
+
+            Err(e) => {
+                error!(error=%e, "failed to fetch actor");
+                Err(e)
+            }
+        }
+    }
+
+    pub async fn follow_actor(&self, actor_uri: &str) -> Result<()> {
+        let Actor { id, inbox, .. } = self.get_actor(actor_uri).await?;
+        info!(%id, %inbox, "sending follow request to inbox");
+
+        let base = base_url();
+        let message_id = Uuid::new_v4();
+        let message = Activity {
+            context: Context::default(),
+            ty: ActivityType::Follow,
+            to: vec![id.clone()],
+            object: IdOrObject::Id(id),
+            id: format!("https://{base}/activities/{message_id}"),
+            actor: format!("https://{base}/actor)"),
+        };
+
+        self.json_post(inbox, message).await?;
+
+        Ok(())
+    }
+
+    pub async fn unfollow_actor(&self, actor_uri: &str) -> Result<()> {
+        let Actor { id, inbox, .. } = self.get_actor(actor_uri).await?;
+        info!(%id, %inbox, "sending unfollow request to inbox");
+
+        let base = base_url();
+        let object_id = Uuid::new_v4();
+        let message_id = Uuid::new_v4();
+        let message = Activity {
+            context: Context::default(),
+            ty: ActivityType::Undo,
+            to: vec![actor_uri.to_owned()],
+            object: IdOrObject::Object {
+                ty: ActivityType::Follow,
+                object: actor_uri.to_owned(),
+                actor: actor_uri.to_owned(),
+                id: format!("https://{base}/activities/{object_id}"),
+            },
+            id: format!("https://{base}/activities/{message_id}"),
+            actor: format!("https://{base}/actor)"),
+        };
+
+        self.json_post(inbox, message).await?;
+
+        Ok(())
+    }
+}
+
+fn map_reqwest_error(uri: impl Into<String>, method: &str, e: reqwest::Error) -> Error {
+    let status = e.status().unwrap_or(StatusCode::INTERNAL_SERVER_ERROR);
+    let error = e.to_string();
+
+    Error::FailedRequest {
+        method: method.to_owned(),
+        status,
+        error,
+        uri: uri.into(),
+    }
+}
+
+#[allow(dead_code)]
+fn new_priv_key() -> RsaPrivateKey {
+    RsaPrivateKey::new(&mut rand::thread_rng(), KEY_LEN).expect("failed to generate a key")
+}
+
+// NOTE: A subset of the rustypub Actor<'a> struct that can implement DeserializeOwned
```

Comment:
if you want me to move rustypub to Owned i'm happy to

Reply:
Yeah I think we'll need that in order to be able to use it for parsing request bodies